flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,734 @@
+                                                    completedCheckpointNum.get() <= 0
+                                                            && pendingMaterialization.stream()
+                                                                    .noneMatch(Future::isDone));
+                                    Thread.sleep(10);
+                                    pendingMaterialization.clear();
+                                    throwArtificialFailure();
+                                } else if (getRuntimeContext().getAttemptNumber() == 1
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitUntilFalse(
+                                            () ->
+                                                    completedCheckpointNum.get() <= 1
+                                                            && pendingMaterialization.stream()
+                                                                    .noneMatch(Future::isDone));
+                                    Thread.sleep(10);
+                                    throwArtificialFailure();
+                                }
+                            }
+                        }));
+    }
+
+    private static AtomicBoolean triggerDelegatedSnapshot = new AtomicBoolean();
+
+    @Test
+    public void testFailedMaterialization() {
+        env.setStateBackend(
+                        new DelegatedStateBackendWrapper(
+                                delegatedStateBackend,
+                                snapshotResultFuture ->
+                                        triggerDelegatedSnapshot.compareAndSet(false, true)
+                                                ? new RunnableFutureWithException<>(
+                                                        snapshotResultFuture)
+                                                : snapshotResultFuture))
+                .setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 0));
+        env.configure(
+                new Configuration()
+                        .set(
+                                StateChangelogOptions.PERIODIC_MATERIALIZATION_INTERVAL,
+                                Duration.ofMillis(20))
+                        .set(StateChangelogOptions.MATERIALIZATION_MAX_FAILURES_ALLOWED, 1));
+        waitAndAssert(
+                buildStreamGraph(
+                        new PreHandlingSource() {
+                            @Override
+                            protected void preHandle() throws Exception {
+                                waitUntilFalse(
+                                        () ->
+                                                currentIndex >= 15
+                                                        && !triggerDelegatedSnapshot.get());
+                            }
+                        }));
+    }
+
+    @Test
+    public void testRescaleOut() throws IOException {
```

Comment:
These rescale cases look like to be better in another IT cases. You could make a abstract Parameterized It case and  make two ITCases extending from it. One is for normal restore, another is for rescaling case.

Reply:
I have splitted it into two ITCases that are extended from `ChangelogPeriodicMaterializationTestBase`.